### PR TITLE
[MXNET-362] ensure same mkldnn engine is used for consistency

### PR DESCRIPTION
**Pull request by ashokei**
## Description ##
Gluon data iterators may trigger different thread for execution context, this causes mkl-dnn engine to be inconsistent. Following snippet reproduces this issue.

```python
import numpy as np
import mxnet as mx
from mxnet import gluon, nd
 
net = gluon.nn.HybridSequential()
with net.name_scope():
    net.add(gluon.nn.Conv2D(channels=32, kernel_size=3, activation=None))
net.collect_params().initialize(mx.init.Xavier(magnitude=2.24), ctx=mx.cpu())
 
val_data = gluon.data.DataLoader(
    gluon.data.vision.CIFAR10(train=False),
    batch_size=32, shuffle=False,num_workers=1)
 
# output should be 0.57521844
X = (32,3,32,32)
y = net(nd.array(np.ones(X))).asnumpy()
print(y[0][0][0][0])
 
# below line works!
# for _ in range(1):
# below line causes bug
for _ in val_data:
    y = net(nd.array(np.ones(X))).asnumpy()
    print(y[0][0][0][0])
    break
```

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 1]
```diff
@@ -67,7 +67,8 @@ class CpuEngine {
  public:
   static CpuEngine *Get() {
     // I's thread-safe in C++11.
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: So we need remove this line of comment?

Comment (ashokei) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h]: comment was correct, it says mkldnn engine is thread safe to use in mkldnn cpp api.

**Comment by ashokei** [comment]
@zheng-da added unittest, @marcoabreu can you please merge if ok. thanks.

File: tests/python/mkl/test_mkldnn.py [hunk 2]
```diff
@@ -89,6 +91,33 @@ def get_tensors(args, shapes, ctx):
...
+    from mxnet import gluon, nd
+
+    net = gluon.nn.HybridSequential()
+    with net.name_scope():
+        net.add(gluon.nn.Conv2D(channels=32, kernel_size=3, activation=None))
+    net.collect_params().initialize(ctx=mx.cpu())
+
+    val_data = gluon.data.DataLoader(
+        gluon.data.vision.CIFAR10(train=False),
+        batch_size=32, shuffle=False, num_workers=1)
```
Comment (zheng-da) [inline tests/python/mkl/test_mkldnn.py]: can you remove this data loader? it doesn't seem this test needs to use it

Comment (ashokei) [reply tests/python/mkl/test_mkldnn.py]: actually, the dataloader is the one that triggers different thread context, i tried without a dataloader (or even if i pass num_workers = 0 to above dataloader) it runs on same thread, so bug wont happen. Not sure if there is any other way to trigger different thread execution context, but this unittest produces the bug currently.

Comment (marcoabreu) [reply tests/python/mkl/test_mkldnn.py]: Thanks a lot for this test! Could you please elaborate the exact behaviour of this unit test in the test with a block comment. I agree with Da that at the moment, it's hard to grasp the exact problem from reading the code. For me it's hard to understand when different threads are getting started and what the exact issue is.

Comment (zheng-da) [reply tests/python/mkl/test_mkldnn.py]: yes, please provide comments why we need data loader here. Thanks

Comment (ashokei) [reply tests/python/mkl/test_mkldnn.py]: Gluon DataLoader allows us to create a new thread as it iterates over data batch. Added comments to PR.

**Comment by ashokei** [comment]
@marcoabreu https://issues.apache.org/jira/browse/MXNET-362

**Comment by ashokei** [comment]
@zheng-da updated with dummy data, can you please accept/review if ok.

**Comment by zheng-da** [comment]
@ashokei it seems when using dummy data, it can't reproduce the bug.
